File: pnix-clj/pnix-clj/src/pnix_clj/io_probe.clj
(ns pnix-clj.io-probe
  (:require [clojure.java.io :as io]
            [pnix-clj.interop :as interop]
            [pnix-clj.json :as json]))

(def lane-classification
  {:lane :core
   :scope :read-only-host-io-probe-adapter
   :role :execute-common-read-only-effect-requests
   :product-runtime :allowed
   :semantic-authority :loads-pnix-meta-owns-no-semantics
   :mutation :read-only-external-filesystem
   :admission :tri-meta-io-gate
   :determinism :required
   :allowed-output :canonical-effect-receipt})

(def capability-class
  {"capability_id" "pnix.io.file-read.v1"
   "entry_point" "host-meta-io-v1"
   "input_signature" "{path:string}"
   "output_shape" "value+receipt"
   "effect_scope" "read-only-filesystem"
   "risk_tier" "bounded-read"
   "discovery_source" "pnix-meta.effect-request.v1"})

(defn request [effect path]
  {"operation_id" effect
   "args" {"path" path}
   "capability_class" capability-class})

(defn probe [root]
  (let [note (str (io/file root "note.txt"))
        missing (str (io/file root "missing.txt"))
        grants #{:file-read}
        exists (interop/apply-effect-request (request "fs.path-exists" note) grants)
        missing-result (interop/apply-effect-request (request "fs.path-exists" missing) grants)
        opened (interop/apply-effect-request (request "fs.open" note) grants)
        typed (interop/apply-effect-request (request "fs.file-type" note) grants)
        listed (interop/apply-effect-request (request "fs.read-dir" root) grants)
        denied (interop/apply-effect-request (request "fs.open" note) #{})
        adapter-error (interop/apply-effect-request (request "fs.open" missing) grants)
        invalid (interop/apply-effect-request (request "fs.open" nil) grants)
        unsupported (interop/apply-effect-request (request "fs.unknown" note) grants)
        report {"schema" "pnix-meta.host-io-probe.v1"
                "adapter_error" (get adapter-error "error")
                "path_exists" (get exists "value")
                "missing_exists" (get missing-result "value")
                "open" (get opened "value")
                "file_type" (get typed "value")
                "read_dir" (get listed "value")
                "denied" (get denied "error")
                "invalid" (get invalid "error")
                "unsupported" (get unsupported "error")
                "receipt_adapter" (get-in opened ["receipt" "adapter"])}]
    (assoc report "all_ok"
           (and (= {"phase" "effect" "class" "effect-adapter-error"}
                   (get report "adapter_error"))
                (= true (get report "path_exists"))
                (= false (get report "missing_exists"))
                (= "hello" (get report "open"))
                (= "regular" (get report "file_type"))
                (= {"note.txt" "regular" "subdir" "directory"}
                   (get report "read_dir"))
                (= {"phase" "effect" "class" "effect-denied"}
                   (get report "denied"))
                (= {"phase" "effect-contract" "class" "invalid-effect-args"}
                   (get report "invalid"))
                (= {"phase" "effect-contract" "class" "unknown-effect-operation"}
                   (get report "unsupported"))
                (= "host-meta-io-v1" (get report "receipt_adapter"))))))

(defn -main [& [root]]
  (let [result (probe root)]
    (println (json/write-json result))
    (when-not (get result "all_ok")
      (System/exit 1))))

File: pnix-hy/pnix-hy/pnix_hy/io_probe.py
"""Cross-host read-only effect adapter probe."""

from __future__ import annotations

import json
from pathlib import Path
import sys

from .interop import apply_effect_request

CAPABILITY_CLASS = {
    "capability_id": "pnix.io.file-read.v1",
    "entry_point": "host-meta-io-v1",
    "input_signature": "{path:string}",
    "output_shape": "value+receipt",
    "effect_scope": "read-only-filesystem",
    "risk_tier": "bounded-read",
    "discovery_source": "pnix-meta.effect-request.v1",
}


def request(effect: str, path: Path) -> dict[str, object]:
    return {
        "operation_id": effect,
        "args": {"path": str(path)},
        "capability_class": CAPABILITY_CLASS,
    }


def probe(root: Path) -> dict[str, object]:
    note = root / "note.txt"
    grants = ("file-read",)
    exists = apply_effect_request(request("fs.path-exists", note), grants)
    missing = apply_effect_request(request("fs.path-exists", root / "missing.txt"), grants)
    opened = apply_effect_request(request("fs.open", note), grants)
    typed = apply_effect_request(request("fs.file-type", note), grants)
    listed = apply_effect_request(request("fs.read-dir", root), grants)
    denied = apply_effect_request(request("fs.open", note), ())
    adapter_error = apply_effect_request(
        request("fs.open", root / "missing.txt"), grants
    )
    invalid_request = request("fs.open", note)
    invalid_request["args"] = {"path": None}
    invalid = apply_effect_request(invalid_request, grants)
    unsupported = apply_effect_request(request("fs.unknown", note), grants)
    report: dict[str, object] = {
        "schema": "pnix-meta.host-io-probe.v1",
        "adapter_error": adapter_error.get("error"),
        "path_exists": exists.get("value"),
        "missing_exists": missing.get("value"),
        "open": opened.get("value"),
        "file_type": typed.get("value"),
        "read_dir": listed.get("value"),
        "denied": denied.get("error"),
        "invalid": invalid.get("error"),
        "unsupported": unsupported.get("error"),
        "receipt_adapter": (opened.get("receipt") or {}).get("adapter"),
    }
    report["all_ok"] = (
        report["adapter_error"]
        == {"phase": "effect", "class": "effect-adapter-error"}
        and report["path_exists"] is True
        and report["missing_exists"] is False
        and report["open"] == "hello"
        and report["file_type"] == "regular"
        and report["read_dir"] == {"note.txt": "regular", "subdir": "directory"}
        and report["denied"] == {"phase": "effect", "class": "effect-denied"}
        and report["invalid"]
        == {"phase": "effect-contract", "class": "invalid-effect-args"}
        and report["unsupported"]
        == {"phase": "effect-contract", "class": "unknown-effect-operation"}
        and report["receipt_adapter"] == "host-meta-io-v1"
    )
    return report


def main() -> int:
    result = probe(Path(sys.argv[1]))
    print(json.dumps(result, sort_keys=True, separators=(",", ":")))
    return 0 if result["all_ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
